rfcs/20210513-pluggable-profiler-for-tensorflow.md

```diff
@@ -0,0 +1,440 @@
+      std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+      profiler_fns_.stop(&profiler_, c_status.get());
+      Status s = tensorflow::StatusFromTF_Status(c_status.get());
+      return s;
+    }
+
+    Status DoCollectData(RunMetadata* run_metadata) {
+      std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+      // Get size of buffer required for Plugin to serialize RunMetadata into.
+      size_t size_in_bytes;
+      profiler_fns_.collect_data_run_metadata(&profiler_, /*buffer=*/nullptr, &size_in_bytes, c_status.get())
+
+      // Prepare an appropriately sized buffer.
+      if (size_in_bytes > 0) {
+        std::vector<uint8_t> buffer(size_in_bytes);
+        profiler_fns_.collect_data_run_metadata(&profiler_, buffer.data(), &size_in_bytes, c_status.get())
+      }
+      return Status::OK();
+    }
+
+    Status DoCollectData(XSpace* space) {
+      std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+      // Get size of buffer required for Plugin to serialize XSpace into.
+      size_t size_in_bytes;
+      profiler_fns_.collect_data_xspace(&profiler_, /*buffer=*/nullptr, &size_in_bytes, c_status.get());
+
+      // Prepare an appropriately sized buffer.
+      if (size_in_bytes > 0) {
+        std::vector<uint8_t> buffer(size_in_bytes);
+        profiler_fns_.collect_data_xspace(&profiler_, buffer.data(), &size_in_bytes, c_status.get());
+        // Deserialize XSpace from the buffer and return it.
+        XSpace plugin_space;
+        plugin_space.ParseFromArray(buffer.data(), buffer.size());
+        for (XPlane& plugin_plane: *plugin_space.mutable_planes()) {
+          XPlane* plane = space->add_planes();
+          plane->Swap(&plugin_plane);
+        }
+      }
+      Status s = tensorflow::StatusFromTF_Status(c_status.get());
+      return s;
+    }
+
+   private:
+     TP_Profiler profiler_;
+     void (*destroy_profiler_)(TP_Profiler*);
+     TP_ProfilerFns profiler_fns_;
+     void (*destroy_profiler_fns_)(TP_ProfilerFns*);
+  };
+
+  ```
+* **PluggableTracer Registration**
+
+  `PluggableTracer` is the implementation of [ProfilerInterface](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/lib/profiler_interface.h#L33), it controls all the registered pluggable profilers. For example, if `device_type` of `ProfileOptions` is set as `PLUGGABLE_DEVICE`, all the registered pluggable profilers will be enabled, when `ProfilerSession` invokes `profiler->Start()`, all the registered pluggable profilers will start the profiling.
+  ```c++
+  class PluggableTracer : public ProfilerInterface {
+   public:
+     explicit PluggableTracer(std::vector<PluginTracerInterface> plugin_interfaces)
+             : plugin_interfaces_(plugin_interfaces) {}
+
+     ~PluggableTracer() override {}
+
+     Status Start() override {
+       for (auto& profiler_interface : plugin_interfaces_) {
+         profiler_interface.DoStart().IgnoreError();
```

Comment:
(Here and below) Errors should not be ignored, and instead should be Update() and returned.

https://abseil.io/docs/cpp/guides/status#keeping-track-of-the-first-error-encountered

Reply:
Thanks for the comments, the consideration here is whether one plugin profiler failed will affect other plugin profiler, I see ProfilerSession, it seems it ignored all the profiler status. 
```
 61 
 62 Status ProfilerSession::CollectData(profiler::XSpace* space) {
 63   mutex_lock l(mutex_);
 64   TF_RETURN_IF_ERROR(status_);
 65 #if !defined(IS_MOBILE_PLATFORM)
 66   LOG(INFO) << "Profiler session collecting data.";
 67   for (auto& profiler : profilers_) {
 68     profiler->Stop().IgnoreError();
 69   }
 70 
 71   for (auto& profiler : profilers_) {
 72     profiler->CollectData(space).IgnoreError();
 73   }
 74 
 75   if (active_) {
 76     // Allow another session to start.
 77     profiler::ReleaseProfilerLock();
 78     active_ = false;
 79   }
 80 
 81   PostProcessSingleHostXSpace(space, start_time_ns_);
 82 #endif
 83 
 84   return Status::OK();
 85 }
```

Reply:
That is a bug in ProfilerSession. The correct behavior should be to `Update()` status for each profiler and not return early upon failure. This way we track the first error encountered and resume calling to other plugin profilers.

Reply:
Done